### DISCUSSION TITLE
Getting Alpha bytes from picture with stream

### DISCUSSION THREAD

**Discussion by Maraxi** [ImageMagick/ImageMagick#2862] [Help] [answered]
I want to stream each channel of my [alpha-test-image](https://user-images.githubusercontent.com/36485759/100164138-87f7f800-2eb7-11eb-8362-4d738bfba2dc.png) to a binary file. I can not get it to work for the alpha channel though; see below:

Getting the Red channel works fine.

    $magick stream -map 'r' alpha-test.png red.dat; hexdump -C red.dat
    00000000  53 06 5b 98 5b 24 24 06  5b 5b 06 06 06 5b 24 24  |S.[.[$$.[[...[$$|
    00000010  24 24 06 5b 06 24 06 06  24 24 24 24 06 5b 24 53  |$$.[.$..$$$$.[$S|
    00000020  06 53 06 24 24 24 06 5b  bf 06 5b 06 06 06 24 24  |.S.$$$.[..[...$$|
    00000030  06 5b 5b 06 06 06 06 06  06 24 06 5b 06 fe 06 06  |.[[......$.[....|
    00000040  bf 24 06 06 06 06 06 fa  5b ad 06 bf 5b 06 53 06  |.$......[...[.S.|
    00000050  fe 5b bf 06 06 bf 06 ad  5b 06 5b 06 06 5b 06 5b  |.[......[.[..[.[|
    00000060  98 06 06 bf                                       |....|
    00000064

Getting the alpha channel however claims every pixel is opaque.

    $magick stream -map 'a' alpha-test.png alpha.dat; hexdump -C alpha.dat
    00000000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
    *
    00000060  ff ff ff ff                                       |....|
    00000064

However the image does contain 45 pixels with Alpha value `x80`

    $convert alpha-test.png -verbose info:
    ...  
    Channel depth:
        Red: 8-bit
        Green: 8-bit
        Blue: 8-bit
        Alpha: 8-bit
    ...
      Channel statistics:
        Alpha:
          min: 128  (0.501961)
          max: 255 (1)
    ...
      Colors: 10
      Histogram:
        45: (6,0,4,128) #06000480 srgba(6,0,4,0.501961)
    ...

    $identify -format '%A' alpha-test.png
    Blend

I can not figure out what my error here is. After all I would expect the byte value `x80` to appear in my `alpha.dat` file.

I am running `ImageMagick 7.0.10-43 Q16 x86_64 2020-11-23` on Arch Linux.

**Comment by snibgo** [comment]
What version of IM, on what platform?

You are doing a hexdump of "pix.dat" but haven't written to that file. I suppose that is a simple copy-paste typo.

This seems to be a bug on IM v7.0.8-64  on Windows 8.1. Alpha always come out as FF, for example:
```
f:\web\im>magick xc:rgba(10%,20%,30%,0) xc:rgba(10%,20%,30%,0.3) +append x.png

f:\web\im>magick stream -map A x.png x.dat

f:\web\im>hexdump -C x.dat
00000000  ff ff                                             |..|
00000002
```
I can't see a problem in stream.c. A developer may take a look.

**Reply by Maraxi** [reply]
Sorry for not renaming my file in the hexdump command. I edited the original post to correct that.
Also added my version.

**Comment by urban-warrior** [comment, marked as answer]
Unfortunately streaming does not work for every image format.  It requires that the image coder read the image pixels once in row order.  PNG does not adhere to the requirement.  To get expected results, try another image format, AVS for example:
```
convert image.png avs:- | magick stream -map 'a' avs:- alpha.dat
```

**Reply by Maraxi** [reply]
Thank you very much for your answer!
It is unfortunately somehow counterintuitive that the color channels can be streamed just fine and alpha does not work. This workaround works just fine though.

**Reply by urban-warrior** [reply]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT master branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://www.imagemagick.org/download/beta/ by sometime tomorrow.
